trunk/core/core.3
.\"	$Csoft: engine.3,v 1.6 2005/06/18 04:25:18 vedge Exp $
.\"
.\" Copyright (c) 2001, 2002, 2003, 2005 CubeSoft Communications, Inc.
.\" <http://www.csoft.org>
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
.\" WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT,
.\" INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
.\" (INCLUDING BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
.\" SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
.\" STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
.\" IN ANY WAY OUT OF THE USE OF THIS SOFTWARE EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.
.\"
.\"	$OpenBSD: mdoc.template,v 1.6 2001/02/03 08:22:44 niklas Exp $
.\"
.Dd NOVEMBER 26, 2001
.Dt AGAR 3
.Os
.ds vT Agar API Reference
.ds oS Agar 1.0
.Sh NAME
.Nm core
.Nd agar core initialization and globals
.Sh SYNOPSIS
.Bd -literal
#include <core/core.h>
.Ed
.Sh DESCRIPTION
.Pp
These functions are responsible for the initialization of Agar's internals,
the initialization of SDL (and possibly OpenGL), the selection of the video
mode, the font engine and the attachment of input devices.
.Pp
.nr nS 1
.Ft "int"
.Fn AG_InitCore "const char *progname" "Uint flags"
.Pp
.Ft "int"
.Fn AG_InitVideo "int w" "int h" "int depth" "Uint flags"
.Pp
.Ft "int"
.Fn AG_InitInput "Uint flags"
.Pp
.Ft "int"
.Fn AG_InitNetwork "Uint flags"
.Pp
.Ft "int"
.Fn AG_InitConfigWin "Uint flags"
.Pp
.Ft "void"
.Fn AG_AtExitFunc "void (*fn)(void)"
.Pp
.Ft "void"
.Fn AG_Quit "void"
.Pp
.Ft "void"
.Fn AG_destroy "void"
.nr nS 0
.Pp
The
.Fn AG_InitCore
function initializes Agar's internal data structures as well as SDL.
It must be called before any other Agar function.
The
.Fa progname
argument specifies the name of the program.
The application's data files will be stored in the
.Pa ~/.<progname>
directory.
This is the only initialization function which is required, the other
.Fn AG_Init*
functions are optional.
.Pp
The
.Fn AG_InitVideo
function sets the video mode, initializes the font engine and GUI.
.Pp
The
.Fn AG_InitInput
function initializes the input devices.
.Pp
The
.Fn AG_InitNetwork
function initializes the network facilities, if the
.Dv NETWORK
option was enabled at compile-time.
.Pp
The
.Fn AG_InitConfigWin
function initializes a dialog window which can be used to modify generic
application settings.
.Pp
The
.Fn AG_AtExitFunc
registers a function that will be invoked automatically by
.Fn AG_Destroy .
.Pp
The
.Fn AG_Quit
function is used to normally terminate the application by pushing an
.Dv SDL_QUIT
event onto the event queue.
.Pp
The
.Fn AG_Destroy
function immediately frees all resources allocated by Agar.
.Sh SEE ALSO
.Xr object 3 ,
.Xr input 3 ,
.Xr map 3 ,
.Xr view 3 ,
.Xr vg 3 ,
.Xr widget 3 ,
.Xr window 3
.Pp
.Bd -literal
http://www.libsdl.org/
http://www.opengl.org/
.Ed


core/core.c
/*	$Csoft: engine.c,v 1.167 2005/10/05 02:00:35 vedge Exp $	*/

/*
 * Copyright (c) 2001, 2002, 2003, 2004, 2005 CubeSoft Communications, Inc.
 * <http://www.csoft.org>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE FOR
 * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
 * USE OF THIS SOFTWARE EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include <config/have_setlocale.h>
#include <config/localedir.h>
#include <config/version.h>
#include <config/network.h>
#include <config/have_pthreads_xopen.h>
#include <config/have_pthread_mutex_recursive.h>
#include <config/have_pthread_mutex_recursive_np.h>
#include <config/have_sdl_cpuinfo.h>

#include <core/core.h>
#include <core/config.h>
#include <core/view.h>
#include <core/typesw.h>

#include <gui/primitive.h>
#include <gui/cursors.h>

#include <stdio.h>
#include <stdlib.h>

#ifdef NETWORK
#include <core/rcs.h>
#endif
#ifdef HAVE_SETLOCALE
#include <locale.h>
#endif
#ifdef HAVE_SDL_CPUINFO
#include <SDL_cpuinfo.h>
#endif

#ifdef THREADS
pthread_mutexattr_t agRecursiveMutexAttr;	/* Recursive mutex attributes */
#endif

const char *agProgName = "";
AG_Config *agConfig;
AG_Object *agWorld;
AG_Object agIconMgr;
void (*agAtexitFunc)(void) = NULL;
AG_Mutex agLinkageLock;
AG_Mutex agTimingLock;
int agServerMode = 0;

int
AG_InitCore(const char *progname, Uint flags)
{
	agProgName = progname;

#ifdef HAVE_SETLOCALE
	setlocale(LC_ALL, "");
#endif
#ifdef ENABLE_NLS
	bindtextdomain("agar", LOCALEDIR);
	textdomain("agar");
#endif

	AG_InitError();

#ifdef THREADS
	pthread_mutexattr_init(&agRecursiveMutexAttr);
#if defined(HAVE_PTHREAD_MUTEX_RECURSIVE_NP)
	pthread_mutexattr_settype(&agRecursiveMutexAttr,
	    PTHREAD_MUTEX_RECURSIVE_NP);
#elif defined(HAVE_PTHREAD_MUTEX_RECURSIVE) || defined(HAVE_PTHREADS_XOPEN)
	pthread_mutexattr_settype(&agRecursiveMutexAttr,
	    PTHREAD_MUTEX_RECURSIVE);
#else
#error "THREADS options requires recursive mutexes"
#endif
	AG_MutexInitRecursive(&agLinkageLock);
	AG_MutexInitRecursive(&agTimingLock);
#endif
	if (SDL_Init(SDL_INIT_TIMER|SDL_INIT_NOPARACHUTE) != 0) {
		AG_SetError("SDL_Init: %s", SDL_GetError());
		return (-1);
	}

	AG_InitTypeSw();
	
	agConfig = Malloc(sizeof(AG_Config), M_OBJECT);
	AG_ConfigInit(agConfig);
	AG_ObjectLoad(agConfig);

	printf("\n");
#ifdef HAVE_SDL_CPUINFO
	printf(_("Architecture extensions: "));
	if (SDL_HasRDTSC()) { printf("RDTSC "); }
	if (SDL_HasMMX()) { printf("MMX "); }
	if (SDL_HasMMXExt()) { printf("MMXExt "); }
	if (SDL_Has3DNow()) { printf("3DNow "); }
	if (SDL_Has3DNowExt()) { printf("3DNowExt "); }
	if (SDL_HasSSE()) { printf("SSE "); }
	if (SDL_HasSSE2()) { printf("SSE2 "); }
	if (SDL_HasAltiVec()) { printf("AltiVec "); }
	printf(".\n");
#endif

	agWorld = AG_ObjectNew(NULL, "world");
	AG_ObjectRemain(agWorld, AG_OBJECT_REMAIN_DATA);
	return (0);
}

int
AG_InitVideo(int w, int h, int bpp, Uint flags)
{
	char path[MAXPATHLEN];
	extern int agBgPopupMenu;
	char *s;

	if (SDL_InitSubSystem(SDL_INIT_VIDEO) != 0) {
		AG_SetError("SDL_INIT_VIDEO: %s", SDL_GetError());
		return (-1);
	}
	SDL_WM_SetCaption(agProgName, agProgName);

	if ((agVideoInfo = SDL_GetVideoInfo()) != NULL) {
		char accel[2048];
		char unaccel[2048];
		size_t size = 2048;

		agVideoFmt = agVideoInfo->vfmt;

		if (agVideoInfo->wm_available) {
			printf(_("Window manager is available.\n"));
		} else {
			printf(_("Window manager is unavailable.\n"));
		}
			
		if (agVideoInfo->hw_available) {
			printf(_("Hardware surfaces are available.\n"));
		} else {
			printf(_("Hardware surfaces are unavailable.\n"));
		}
		
		accel[0] = '\0';
		unaccel[0] = '\0';

		strlcat(agVideoInfo->blit_hw ? accel : unaccel,
		    _("\tSDL hardware blits\n"), size);
		strlcat(agVideoInfo->blit_hw_CC ? accel : unaccel,
		    _("\tSDL hardware->hardware colorkey blits\n"), size);
		strlcat(agVideoInfo->blit_hw_A ? accel : unaccel,
		    _("\tSDL hardware->hardware alpha blits\n"), size);
		strlcat(agVideoInfo->blit_sw ? accel : unaccel,
		    _("\tSDL software->hardware blits\n"), size);
		strlcat(agVideoInfo->blit_sw_CC ? accel : unaccel,
		    _("\tSDL software->hardware colorkey blits\n"), size);
		strlcat(agVideoInfo->blit_sw_A ? accel : unaccel,
		    _("\tSDL software->hardware alpha blits\n"), size);
		strlcat(agVideoInfo->blit_fill ? accel : unaccel,
		    _("\tSDL color fills\n"), size);

		if (accel[0] != '\0')
			printf(_("Accelerated operations:\n%s"), accel);
		if (unaccel[0] != '\0')
			printf(_("Unaccelerated operations:\n%s"), unaccel);

		printf("\n");
	}

	if (flags & (AG_VIDEO_OPENGL|AG_VIDEO_OPENGL_OR_SDL)) {
#ifdef HAVE_OPENGL
		AG_SetBool(agConfig, "view.opengl", 1);
#else
		if ((flags & AG_VIDEO_OPENGL_OR_SDL) == 0)
			fatal("Agar OpenGL support is not compiled in");
#endif
	}
	if (AG_ViewInit(w, h, bpp, flags) == -1 || AG_TextInit() == -1) {
		return (-1);
	}
	AG_ColorsInit();
	AG_InitPrimitives();
	AG_CursorsInit();

	strlcpy(path, AG_String(agConfig, "save-path"), sizeof(path));
	strlcat(path, AG_PATHSEP, sizeof(path));
	strlcat(path, "gui-colors.acs", sizeof(path));
	(void)AG_ColorsLoad(path);

#ifdef HAVE_OPENGL
	if (agView->opengl) {
		Uint8 r, g, b;

		SDL_GetRGB(AG_COLOR(BG_COLOR), agVideoFmt, &r, &g, &b);
		glClearColor(r/255.0, g/255.0, b/255.0, 1.0);
	} else
#endif
	{
		SDL_FillRect(agView->v, NULL, AG_COLOR(BG_COLOR));
		SDL_UpdateRect(agView->v, 0, 0, agView->w, agView->h);
	}

	if (flags & AG_VIDEO_BGPOPUPMENU) { agBgPopupMenu = 1; }
	
	AG_ObjectInit(&agIconMgr, "object", "IconMgr", NULL);
	if (AG_WireGfx(&agIconMgr, "core-icons") == -1) {
		fatal("icons: %s", AG_GetError());
	}
	return (0);
}

int
AG_InitConfigWin(Uint flags)
{
	AG_ConfigWindow(agConfig, flags);
	return (0);
}

/* Initialize graphics and input devices. */
int
AG_InitInput(Uint flags)
{
	extern int agKbdUnicode;			/* config.c */
	int i, n, njoys;

	SDL_EnableUNICODE(agKbdUnicode || (flags & AG_FORCE_UNICODE));

	if (AG_Bool(agConfig, "input.joysticks") ||
	    (flags & AG_FORCE_JOYSTICK)) {
		if (SDL_InitSubSystem(SDL_INIT_JOYSTICK) == 0) {
			n = SDL_NumJoysticks();
			for (i = 0, njoys = 0; i < n; i++) {
				if (SDL_JoystickOpen(i) != NULL)
					njoys++;
			}
			if (njoys > 0)
				SDL_JoystickEventState(SDL_ENABLE);
		}
	}
	return (0);
}

int
AG_InitNetwork(Uint flags)
{
#if defined(DEBUG) && defined(NETWORK) && defined(THREADS)
	if ((flags & AG_INIT_DEBUG_SERVER) && agServerMode) {
		extern int AG_DebugServerStart(void);
		AG_DebugServerStart();
	}
#endif
#if defined(NETWORK)
	if (flags & AG_INIT_RCS)
		AG_RcsInit();
#endif
	return (0);
}

/* Register a function to invoke in AG_Quit(). */
void
AG_AtExitFunc(void (*func)(void))
{
	agAtexitFunc = func;
}

/* Request a graceful shutdown of the application. */
void
AG_Quit(void)
{
	SDL_Event nev;

	nev.type = SDL_QUIT;
	SDL_PushEvent(&nev);
}

/* Immediately clean up and exit the application. */
void
AG_Destroy(void)
{
	if (agAtexitFunc != NULL)
		agAtexitFunc();

#ifdef NETWORK
	AG_RcsDestroy();
#endif

	AG_ObjectDestroy(agWorld);
	AG_TextDestroy();
	AG_ViewDestroy();
	
	AG_ColorsDestroy();
	AG_CursorsDestroy();

	AG_ObjectDestroy(agConfig);
	Free(agConfig, M_OBJECT);

/*	AG_MutexDestroy(&agLinkageLock); */
	AG_MutexDestroy(&agTimingLock);

	AG_DestroyError();
	AG_DestroyTypeSw();
	SDL_Quit();
	exit(0);
}

